router/http.go

```diff
@@ -368,101 +395,122 @@ func (h *httpSyncHandler) Remove(id string) error {
+		port, _ := strconv.Atoi(mustPortFromAddr(addr))
+		certForHandshake := func(hello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			r := s.findRoute(hello.ServerName, port, "/")
+			if r == nil {
+				return nil, errMissingTLS
+			}
+			return r.keypair, nil
 		}
-		return r.keypair, nil
-	}
-	tlsConfig := tlsconfig.SecureCiphers(&tls.Config{
-		GetCertificate: certForHandshake,
-		Certificates:   []tls.Certificate{s.keypair},
-		NextProtos:     []string{http2.NextProtoTLS, "h2-14"},
-	})
+		tlsConfig := tlsconfig.SecureCiphers(&tls.Config{
+			GetCertificate: certForHandshake,
+			Certificates:   []tls.Certificate{s.keypair},
+			NextProtos:     []string{http2.NextProtoTLS, "h2-14"},
+		})
 
-	l, err := listenFunc("tcp4", s.TLSAddr)
-	if err != nil {
-		return listenErr{s.Addr, err}
-	}
-	if s.proxyProtocol {
-		l = proxyproto.Listener{l}
-	}
-	s.tlsListener = tls.NewListener(l, tlsConfig)
+		l, err := listenFunc("tcp4", addr)
+		if err != nil {
+			return listenErr{addr, err}
+		}
+		if s.proxyProtocol {
+			l = proxyproto.Listener{l}
+		}
+		listener := tls.NewListener(l, tlsConfig)
+		s.tlsListeners = append(s.tlsListeners, listener)
 
-	handler := fwdProtoHandler{
-		Handler: s,
-		Proto:   "https",
-		Port:    mustPortFromAddr(s.tlsListener.Addr().String()),
-	}
-	http2Server := &http2.Server{}
-	http2Handler := func(hs *http.Server, c *tls.Conn, h http.Handler) {
-		http2Server.ServeConn(c, &http2.ServeConnOpts{
-			Handler:    handler,
-			BaseConfig: hs,
-		})
-	}
+		handler := fwdProtoHandler{
+			Handler: s,
+			Proto:   "https",
+			Port:    mustPortFromAddr(listener.Addr().String()),
+		}
+
+		http2Server := &http2.Server{}
+		http2Handler := func(hs *http.Server, c *tls.Conn, h http.Handler) {
+			http2Server.ServeConn(c, &http2.ServeConnOpts{
+				Handler:    handler,
+				BaseConfig: hs,
+			})
+		}
 
-	server := &http.Server{
-		Addr:    s.tlsListener.Addr().String(),
-		Handler: handler,
-		TLSNextProto: map[string]func(*http.Server, *tls.Conn, http.Handler){
-			http2.NextProtoTLS: http2Handler,
-			"h2-14":            http2Handler,
-		},
+		server := &http.Server{
+			Addr:    listener.Addr().String(),
+			Handler: handler,
+			TLSNextProto: map[string]func(*http.Server, *tls.Conn, http.Handler){
+				http2.NextProtoTLS: http2Handler,
+				"h2-14":            http2Handler,
+			},
+		}
+
+		// TODO: log error
+		go server.Serve(listener)
 	}
 
-	// TODO: log error
-	go server.Serve(s.tlsListener)
 	return nil
 }
 
-func (s *HTTPListener) findRoute(host string, path string) *httpRoute {
+func (s *HTTPListener) findRoute(host string, portInt int, path string) *httpRoute {
 	host = strings.ToLower(host)
 	if strings.Contains(host, ":") {
 		host, _, _ = net.SplitHostPort(host)
 	}
+	port := strconv.Itoa(portInt)
+	addressArrays := [][]string{s.Addrs, s.TLSAddrs}
+	for _, addressArray := range addressArrays {
+		if len(addressArray) > 0 {
+			_, defaultPort, _ := net.SplitHostPort(addressArray[0])
```

Comment:
Hrm, okay, you're probably right. I'm actually doing a lot of hoop jumping for just a little savings in data structure. Changing that up.